### PR TITLE
Add support for Audio and Video summarization to Docsum

### PR DESCRIPTION
**Pull request by MSCetin37**
## Description

Extend the current Document Summarization Application by incorporating video and audio summary features. This enhancement will enable the application to summarize video and audio content in addition to text documents, thereby broadening its utility and applicability.

## Issues
https://github.com/opea-project/docs/blob/main/community/rfcs/24-06-21-OPEA-001-DocSum_Video_Audio.md

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [x] Breaking change (fix or feature that would break existing design and interface)
- [ ] Others (enhancement, documentation, validation, etc.)

### Whisper Service

Run the following command to validate the Whisper Service:

```bash
python comps/asr/whisper/dependency/check_whisper_server.py 
```

Expected output:

```
{'asr_result': 'who is pat gelsinger'}
```

### Audio2Text Service

Run the following command to validate the Audio2Text Service:

```bash
python comps/dataprep/multimedia2text/audio2text/check_a2t_server.py
```

Expected output:

```
{'downstream_black_list': [], 'id': '21b0459477abea6d85d20f4b5ddcb714', 'query': 'who is pat gelsinger'}
```

*Note: The `id` value will be different.*

### Video2Audio Service

Run the following command to validate the Video2Audio Service:

```bash
python comps/dataprep/multimedia2text/video2audio/check_v2a_microserver.py
```

Expected output:

```
========= Audio file saved as ======
comps/dataprep/multimedia2text/video2audio/converted_audio.wav
====================================
```

### Multimedia2Text Service

Run the following command to validate the Multimedia2Text Service:

```bash
python comps/dataprep/multimedia2text/check_multimedia2text.py 
```

Expected output:

```
Running test: Whisper service
>>> Whisper service Test Passed ... 

Running test: Audio2Text service
>>> Audio2Text service Test Passed ... 

Running test: Video2Text service
>>> Video2Text service Test Passed ... 

Running test: Multimedia2text service
>>> Multimedia2text service test for text data type passed ... 
>>> Multimedia2text service test for audio data type passed ... 
>>> Multimedia2text service test for video data type passed ... 
```

```bash
cd GenAIComps/tests
bash tests/dataprep/test_dataprep_multimedia.sh
```

### CODE REVIEW

File: comps/dataprep/multimedia2text/video2audio/video2audio_microservice.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import json
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/video2audio/video2audio_microservice.py:1]: Please add copyright header

**Review by ashahba — CHANGES_REQUESTED** [review]
Thanks @MSCetin37 and @HarshaRamayanam

File: comps/asr/whisper/dependency/whisper_model.py [hunk 2]
```diff
@@ -14,7 +14,14 @@
...
     """Convert audio to text."""
 
-    def __init__(self, model_name_or_path="openai/whisper-small", language="english", device="cpu", hpu_max_len=8192):
+    def __init__(
+        self,
+        model_name_or_path="openai/whisper-small",
+        language="english",
+        device="cpu",
+        hpu_max_len=8192,
+        return_timestamps=False,
```
[the remark this answers is not in the document]
Comment (MSCetin37) [reply comps/asr/whisper/dependency/whisper_model.py]: it is set to false due to Backward Compatibility

File: .github/workflows/docker/compose/dataprep-compose-cd.yaml [hunk 3]
```diff
@@ -31,3 +31,15 @@ services:
...
+      dockerfile: comps/dataprep/multimedia2text/Dockerfile
+    image: ${REGISTRY:-opea}/dataprep-multimedia2text:${TAG:-latest}
+  dataprep-video2audio:
+    build:
+      dockerfile: comps/dataprep/multimedia2text/video2audio/Dockerfile
+    image: ${REGISTRY:-opea}/dataprep-video2audio:${TAG:-latest}
+  dataprep-audio2text:
+    build:
+      dockerfile: comps/dataprep/multimedia2text/video2audio/Dockerfile
+    image: ${REGISTRY:-opea}/dataprep-audio2text:${TAG:-latest}
```
Comment (ashahba) [inline .github/workflows/docker/compose/dataprep-compose-cd.yaml:45]: End of file char missing

File: comps/dataprep/multimedia2text/README.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+   To remove all unused images, you can use:
+
+   ```bash
+   docker image prune -a
+   ```
+
+
+
+
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/README.md:224]: Please remove these lines

File: comps/dataprep/multimedia2text/audio2text/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    fi
+
+# Set the PYTHONPATH environment variable
+ENV PYTHONPATH=$PYTHONPATH:/home/user
+
+# Set the working directory
+WORKDIR /home/user/comps/dataprep/multimedia2text/audio2text
+
+# Define the entry point for the container
+ENTRYPOINT ["python", "audio2text.py"]
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/audio2text/Dockerfile:37]: End of file char missing

File: comps/dataprep/multimedia2text/audio2text/audio2text.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        
+        # Log initialization message
+        logger.info("[a2t - router] A2T initialized.")
+        
+        # Start the microservice
+        opea_microservices["opea_service@a2t"].start()
+        
+    except Exception as e:
+        logger.error(f"Failed to start the microservice: {e}")
+        raise
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/audio2text/audio2text.py:88]: End of file char missing

File: comps/dataprep/multimedia2text/audio2text/check_a2t_server.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+# Define the endpoint and the input data
+endpoint = "http://localhost:9099/v1/audio/transcriptions"
+inputs = {"byte_str": test_audio_base64_str}
+
+# Send the POST request to the endpoint
+response = requests.post(url=endpoint, data=json.dumps(inputs), proxies={"http": None})
+
+# Print the response from the server
+print(response.json())
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/audio2text/check_a2t_server.py:33]: End of file char missing.
Also print statements are used mostly for debugging.
If this has no use other than debugging, I'd remove it.

File: comps/dataprep/multimedia2text/check_multimedia2text.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+if __name__ == "__main__":
+    # Run the tests and print the results
+    try:
+        test_whisper_service()
+        test_audio2text()
+        test_video2text()
+        test_multimedia2text_data()
+        
+    except AssertionError as e:
+        print(f"Test failed: {e}")
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/check_multimedia2text.py:145]: End of file char missing.

File: comps/dataprep/multimedia2text/multimedia2text.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+        # Log initialization message
+        logger.info("[multimedia2text - router] multimedia2text initialized.")
+
+        # Start the microservice
+        opea_microservices["opea_service@multimedia2text"].start()
+
+    except Exception as e:
+        logger.error(f"Failed to start the multimedia2text microservice: {e}")
+        raise
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/multimedia2text.py:96]: End of file char missing.

File: comps/dataprep/multimedia2text/video2audio/video2audio.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        uid = str(uuid.uuid4())
+        audio_file_name = uid + ".mp3"
+        
+        # Convert the video to audio
+        self.convert_video_to_audio(video_file_name, audio_file_name)
+        
+        # Convert the audio file to a base64 encoded string
+        base64_str = self.convert_base64(audio_file_name)
+        
+        return base64_str
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/video2audio/video2audio.py:90]: End of file char missing.

File: comps/dataprep/multimedia2text/video2audio/video2audio_microservice.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+        # Log initialization message
+        logger.info("[video2audio - router] VIDEO2AUDIO initialized.")
+        
+        # Start the microservice
+        opea_microservices["opea_service@video2audio"].start()
+        
+    except Exception as e:
+        logger.error(f"Failed to start the microservice: {e}")
+        raise
```
Comment (ashahba) [inline comps/dataprep/multimedia2text/video2audio/video2audio_microservice.py:87]: End of file char missing.

**Review by ashahba — CHANGES_REQUESTED** [review]

**Review by ashahba — APPROVED** [review]
LGTM!

**Review by mkbhanda — APPROVED** [review]
LGTM

**Comment by lvliang-intel** [comment]
@MSCetin37,
please update comps/dataprep/multimedia2text/audio2text/Dockerfile to .github/workflows/docker/compose/dataprep-compose.yaml. The yaml is used for release images build.

**Comment by lianhao** [comment]
The audio2text service is almost the same as the [asr service](https://github.com/opea-project/GenAIComps/blob/main/comps/asr/whisper/asr.py) except the returned data format. Why can't we combine these 2 services into one?

**Pull request merged**